pyron-plugin/src/main/scala/com/cloudentity/pyron/plugin/util/value/ValueResolver.scala

```diff
@@ -1,54 +1,104 @@
 package com.cloudentity.pyron.plugin.util.value
 
 import com.cloudentity.pyron.domain.flow.{AuthnCtx, RequestCtx}
+import com.cloudentity.pyron.plugin.util.value.PatternUtil.safePatternAndParams
 import io.circe.Json
 import io.vertx.core.json.{JsonArray, JsonObject}
 
 import java.util.{List => JavaList, Map => JavaMap}
 import scala.annotation.tailrec
+import scala.collection.JavaConverters._
 import scala.util.Try
-import scala.util.matching.Regex
-
 
 object ValueResolver extends ValueResolver
 trait ValueResolver {
+
   def resolveString(req: RequestCtx, bodyOpt: Option[JsonObject], valueOrRef: ValueOrRef): Option[String] =
-    valueOrRef match {
-      case Value(value)        => value.asString
-      case BodyRef(path)       => bodyOpt.flatMap(extractBodyAttribute(_, path)).flatMap(_.asString)
-      case PathParamRef(param) => req.request.uri.pathParams.value.get(param)
-      case AuthnRef(path)      => extractAuthnCtxAttribute(req.authnCtx, path).flatMap(_.asString)
-      case HeaderRef(header, _) => req.request.headers.get(header) // always take first header value
+    (valueOrRef: @unchecked) match { // no default case, fail fast
+      case Value(value) => value.asString
+      case HostRef => Option(req.original.host).filter(_.nonEmpty)
+      case HostNameRef => resolveHostName(req)
+      case HostPortRef => resolveHostPort(req)
+      case SchemeRef => resolveScheme(req)
+      case LocalHostRef => resolveLocalHost(req)
+      case RemoteHostRef => resolveRemoteHost(req)
+      case CookieRef(cookie) => resolveCookie(req, cookie)
+      case BodyRef(path) => resolveBody(bodyOpt, path).flatMap(_.asString)
+      case PathParamRef(param) => resolvePathParam(req, param)
+      case QueryParamRef(param) => req.request.uri.query.get(param) // take first param value
+      case HeaderRef(header, AllHeaderRefType) => req.request.headers.get(header) // take first header value
+      case HeaderRef(header, FirstHeaderRefType) => req.request.headers.get(header) // take first header value
+      case AuthnRef(path) => extractAuthnCtxAttribute(req.authnCtx, path).flatMap(_.asString)
     }
 
-  def resolveString(req: RequestCtx, valueOrRef: ValueOrRef): Option[String] =
-    valueOrRef match {
-      case BodyRef(_) => resolveString(req, req.request.bodyOpt.flatMap(buf => Try(buf.toJsonObject).toOption), valueOrRef)
-      case x          => resolveString(req, None, x)
-    }
+  def resolveString(req: RequestCtx, valueOrRef: ValueOrRef): Option[String] = valueOrRef match {
+    case BodyRef(_) => resolveString(req, req.request.bodyOpt.flatMap(buf => Try(buf.toJsonObject).toOption), valueOrRef)
+    case x          => resolveString(req, None, x)
+  }
 
   def resolveListOfStrings(req: RequestCtx, bodyOpt: Option[JsonObject], valueOrRef: ValueOrRef): Option[List[String]] =
-    valueOrRef match {
+    (valueOrRef: @unchecked) match { // no default case, fail fast
```

Comment:
Let's use exhaustive pattern matching and this way not allow throwing exceptions at runtime.

Reply:
but, there was this notion of doing quick fail when reading config. It was originally unchecked (no '_' case), I think in previous review you warned me about not adding '_' since it would delay failure, I merely kept it that way.

Reply:
What do you mean by 'fail fast' in this case? Maybe I misunderstood, but I thought it may fail at runtime (combined with unchecked annotation). IMO it should fail at compile time and most likely it would as ValueOrRef is a sealed trait, is it so? But what is the reason to have `unchecked` here?

Reply:
What do you mean by 'fail fast' in this case? Maybe I misunderstood, but I thought it may fail at runtime (combined with unchecked annotation). IMO it should fail at compile time and most likely it would as ValueOrRef is a sealed trait, is it so? But what is the reason to have `unchecked` here?

Reply:
Oops, sorry for confusion, IntelliJ tricked me when I added new RefType and did not fail compilation. After second build it failed. Actually, there is no need to rework the pattern matching to fail at non-exhaustive pattern matching, it would fail anyway, or it wouldn't? :)

Reply:
done